scrapy/extensions/periodic_log.py

```diff
@@ -15,52 +25,59 @@ class PeriodicLog:
+        ext_delta: Dict[str, Any] = {},
+        ext_timing_enabled: bool = False,
     ):
-        self.stats = stats
-        self.interval = interval
-        self.multiplier = 60.0 / self.interval
-        self.task = None
-        self.encoder = ScrapyJSONEncoder(sort_keys=True, indent=4)
-        self.ext_stats_enabled = bool(ext_stats)
-        self.ext_stats_include = ext_stats.get("include", [])
-        self.ext_stats_exclude = ext_stats.get("exclude", [])
-        self.ext_delta_enabled = bool(ext_delta)
-        self.ext_delta_include = ext_delta.get("include", [])
-        self.ext_delta_exclude = ext_delta.get("exclude", [])
-        self.ext_timing_enabled = ext_timing_enabled
+        self.stats: StatsCollector = stats
+        self.interval: float = interval
+        self.multiplier: float = 60.0 / self.interval
+        self.task: Optional[task.LoopingCall] = None
+        self.encoder: JSONEncoder = ScrapyJSONEncoder(sort_keys=True, indent=4)
+        self.ext_stats_enabled: bool = bool(ext_stats)
+        self.ext_stats_include: List[str] = ext_stats.get("include", [])
+        self.ext_stats_exclude: List[str] = ext_stats.get("exclude", [])
+        self.ext_delta_enabled: bool = bool(ext_delta)
+        self.ext_delta_include: List[str] = ext_delta.get("include", [])
+        self.ext_delta_exclude: List[str] = ext_delta.get("exclude", [])
+        self.ext_timing_enabled: bool = ext_timing_enabled
 
     @classmethod
-    def from_crawler(cls, crawler):
-        interval = crawler.settings.getfloat("LOGSTATS_INTERVAL")
+    def from_crawler(cls, crawler: Crawler) -> Self:
+        interval: float = crawler.settings.getfloat("LOGSTATS_INTERVAL")
         if not interval:
             raise NotConfigured
         try:
-            ext_stats = crawler.settings.getdict("PERIODIC_LOG_STATS")
+            ext_stats: Optional[Dict[str, Any]] = crawler.settings.getdict(
+                "PERIODIC_LOG_STATS"
+            )
         except (TypeError, ValueError):
             ext_stats = (
                 {"enabled": True}
                 if crawler.settings.getbool("PERIODIC_LOG_STATS")
                 else None
             )
         try:
-            ext_delta = crawler.settings.getdict("PERIODIC_LOG_DELTA")
+            ext_delta: Optional[Dict[str, Any]] = crawler.settings.getdict(
+                "PERIODIC_LOG_DELTA"
+            )
         except (TypeError, ValueError):
             ext_delta = (
                 {"enabled": True}
                 if crawler.settings.getbool("PERIODIC_LOG_DELTA")
                 else None
             )
 
-        ext_timing_enabled = crawler.settings.getbool(
+        ext_timing_enabled: bool = crawler.settings.getbool(
             "PERIODIC_LOG_TIMING_ENABLED", False
         )
         if not (ext_stats or ext_delta or ext_timing_enabled):
             raise NotConfigured
+        assert crawler.stats
+        assert ext_stats is not None
+        assert ext_delta is not None
```

Comment:
The logic for these two settings is complicated because they can (per the docs) be `None` (which is equivalent to an empty dict after `getdict()` anyway), `True` or a dict and so can't be False. Setting one of them to False indeed breaks the code. So the code above that sets the vars to None if getbool() returned False is dead code that can be refactored.